File: src/main/java/de/simpletactics/wiki/lib/adapter/WikiAdapter.java
package de.simpletactics.wiki.lib.adapter;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import de.simpletactics.wiki.lib.model.WikiType;
import de.simpletactics.wiki.lib.service.port.WikiEntryPort;
import de.simpletactics.wiki.lib.service.port.WikiPort;
import de.simpletactics.wiki.lib.service.port.WikiRightsPort;
import de.simpletactics.wiki.lib.util.WikiMapper;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import javax.annotation.PostConstruct;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Component;

@Component
@AllArgsConstructor
@SuppressWarnings("unchecked")
public class WikiAdapter implements WikiPort {

  private WikiEntryPort wikiEntryPort;
  private WikiRightsPort wikiRightsPort;
  private Gson gson;

  @PostConstruct
  public void initGsonObject() {
    GsonBuilder builder = new GsonBuilder();
    builder.setPrettyPrinting();
    this.gson = builder.create();
  }

  public String getTopicsAsJson(int id, String username) {

    if (!wikiRightsPort.getWikiBerechtigung(username, id).equals("none")) {
      if (id == 0 || getWikiType(id) == WikiType.THEMENBEREICH) {

        List<Map<String, Object>> filteredTopics = getFilteredTopics(id, username);
        List<Map<String, Object>> filteredContents = getFilteredContents(id, username);

        if (id == 0) {
          return gson
              .toJson(WikiMapper.subjectEntryMapper(id,
                  wikiRightsPort.getWikiBerechtigung(username, id), filteredTopics,
                  filteredContents, List.of(), wikiEntryPort.getTopic(id),
                  getWikiType(id), List.of()));
        } else {
          return gson
              .toJson(WikiMapper.subjectEntryMapper(id,
                  wikiRightsPort.getWikiBerechtigung(username, id), filteredTopics,
                  filteredContents, getFilteredNavTopics(id, username), wikiEntryPort.getTopic(id),
                  getWikiType(id), getFilteredNavContents(id, username)));
        }


      } else if (getWikiType(id) == WikiType.STANDARDEINTRAG) {
        List<Map<String, Object>> content = wikiEntryPort.getContent(id);
        return gson.toJson(
            WikiMapper.wikiEntryMapper(wikiRightsPort.getWikiBerechtigung(username, id), content,
                getFilteredNavTopics(id, username), wikiEntryPort.getTopic(id), getWikiType(id),
                getFilteredNavContents(id, username)));
      }
    }
    return "access denied";
  }


  public <T> T getTopicsAsObject(int id, String username) {

    if (!wikiRightsPort.getWikiBerechtigung(username, id).equals("none")) {
      if (id == 0 || getWikiType(id) == WikiType.THEMENBEREICH) {

        List<Map<String, Object>> filteredTopics = getFilteredTopics(id, username);
        List<Map<String, Object>> filteredContents = getFilteredContents(id, username);

        return (T) WikiMapper.subjectEntryMapper(id,
            wikiRightsPort.getWikiBerechtigung(username, id), filteredTopics,
            filteredContents, getFilteredNavTopics(id, username), wikiEntryPort.getTopic(id),
            getWikiType(id), getFilteredNavContents(id, username));

      } else if (getWikiType(id) == WikiType.STANDARDEINTRAG) {
        List<Map<String, Object>> content = wikiEntryPort.getContent(id);
        return (T) WikiMapper.wikiEntryMapper(wikiRightsPort.getWikiBerechtigung(username, id),
            content,
            getFilteredNavTopics(id, username), wikiEntryPort.getTopic(id), getWikiType(id),
            getFilteredNavContents(id, username));
      }
    }
    return null;
  }

  public void updateEntry(String topic, String content, int id) {
    wikiEntryPort.updateContent(topic, content, id);
  }

  public void updateTopic(String topic, int id) {
    wikiEntryPort.updateTopic(topic, id);
  }

  public int addEntry(String topic, String content, String threadId, WikiType wikiType) {
    int newId = wikiEntryPort.newId(wikiType == WikiType.STANDARDEINTRAG ? "2" : "3");
    wikiEntryPort.linkIds(Integer.parseInt(threadId), newId);
    wikiEntryPort.updateContent(topic, content, newId);
    wikiRightsPort.setWikiRightsCopyOvertopic(newId, Integer.parseInt(threadId));
    return newId;
  }

  public int addTopic(String id, String topic) {
    int newId = wikiEntryPort.newId("1");
    wikiEntryPort.linkIds(Integer.parseInt(id), newId);
    wikiEntryPort.updateTopic(topic, newId);
    wikiRightsPort.setWikiRightsCopyOvertopic(newId, Integer.parseInt(id));
    return newId;
  }

  public void deleteTopic(int id) {
    wikiEntryPort.deleteTopic(id);
  }

  public void deleteEntry(int id) {
    wikiEntryPort.deleteContent(id);
  }

  public WikiType getWikiType(int id) {
    List<Map<String, Object>> type = wikiEntryPort.getContent(id);
    switch (Integer.parseInt(type.get(0).get("type").toString())) {
      case 1:
        return WikiType.THEMENBEREICH;
      case 2:
        return WikiType.STANDARDEINTRAG;
      case 3:
        return WikiType.POLL;
      default:
        return null;
    }

  }

  private List<Map<String, Object>> getFilteredTopics(int id, String username) {
    List<Map<String, Object>> topics = new ArrayList<>();

    List<Map<String, Object>> returnedTopics;
    returnedTopics = wikiEntryPort.getTopics(
        id); //Alle Topics mit id-Verknüpfung in der Datenbank werden in Liste geladen

    //filtert die Einträge aus rTopics, die gelesen werden dürfen und trägt diese in topics ein
    for (Map<String, Object> rTopic : returnedTopics) {
      String rights = wikiRightsPort.getWikiBerechtigung(username,
          Integer.parseInt(rTopic.get("id").toString()));
      if (!rights.equals("none")) {
        topics.add(rTopic);
      }
    }

    return topics;
  }

  private List<Map<String, Object>> getFilteredContents(int id, String username) {
    List<Map<String, Object>> contents = new ArrayList<>();

    List<Map<String, Object>> returnedContents;
    returnedContents = wikiEntryPort.getContents(
        id); //Alle Contents mit id-Verknüpfung in der Datenbank werden in Liste geladen

    //filtert die Einträge aus rContents, die gelesen werden dürfen und trägt diese in contents ein
    for (Map<String, Object> rContent : returnedContents) {
      String rights = wikiRightsPort.getWikiBerechtigung(username,
          Integer.parseInt(rContent.get("id").toString()));
      if (!rights.equals("none")) {
        contents.add(rContent);
      }
    }

    return contents;
  }

  private List<Map<String, Object>> getFilteredNavTopics(int id, String username) {

    List<Map<String, Object>> parentSubject = wikiEntryPort.getTopic(id);
    List<Map<String, Object>> topics = new ArrayList<>();

    List<Map<String, Object>> returnedTopics;
    returnedTopics = wikiEntryPort.getTopics(Integer.parseInt(parentSubject.get(0).get("id")
        .toString())); //Alle Topics mit id-Verknüpfung in der Datenbank werden in Liste geladen

    //filtert die Einträge aus rTopics, die gelesen werden dürfen und trägt diese in topics ein
    for (Map<String, Object> rTopic : returnedTopics) {
      String rights = wikiRightsPort.getWikiBerechtigung(username,
          Integer.parseInt(rTopic.get("id").toString()));
      if (!rights.equals("none")) {
        topics.add(rTopic);
      }
    }

    return topics;
  }

  private List<Map<String, Object>> getFilteredNavContents(int id, String username) {

    List<Map<String, Object>> parentSubject = wikiEntryPort.getTopic(id);
    List<Map<String, Object>> contents = new ArrayList<>();
    int subjectId = Integer.parseInt(parentSubject.get(0).get("id").toString());
    List<Map<String, Object>> returnedContents;

    returnedContents = wikiEntryPort.getContents(
        subjectId); //Alle Contents mit id-Verknüpfung in der Datenbank werden in Liste geladen

    //filtert die Einträge aus rContents, die gelesen werden dürfen und trägt diese in contents ein
    for (Map<String, Object> rContent : returnedContents) {
      String rights = wikiRightsPort
          .getWikiBerechtigung(username, Integer.parseInt(rContent.get("id").toString()));
      if (!rights.equals("none")) {
        contents.add(rContent);
      }
    }

    return contents;
  }

}

File: src/main/kotlin/de/simpletactics/wiki/lib/adapter/WikiAdapter.kt
package de.simpletactics.wiki.lib.adapter

import de.simpletactics.wiki.lib.adapter.dto.EntryEntity
import de.simpletactics.wiki.lib.adapter.dto.TopicEntity
import de.simpletactics.wiki.lib.adapter.persistence.WikiSqlAdapter
import de.simpletactics.wiki.lib.model.WikiType
import de.simpletactics.wiki.lib.services.port.WikiPort
import de.simpletactics.wiki.lib.util.checkAccess
import de.simpletactics.wiki.lib.util.verify
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
class WikiAdapter(
    private val wikiSqlAdapter: WikiSqlAdapter,
) : WikiPort {

    override fun getTopic(id: Int, hasAccess: () -> Boolean): TopicEntity? {
        hasAccess.checkAccess("Access denied to get topic for id $id")
        return wikiSqlAdapter.getTopic(id)
    }

    @Transactional
    override fun createTopic(parentId: Int, topic: String, hasAccess: () -> Boolean): Int {
        hasAccess.checkAccess("Access denied for creating topic for parentId $parentId")

        val wikiType = wikiSqlAdapter.getWikiType(parentId)
        val parent = wikiSqlAdapter.getTopic(parentId)

        verify(
            wikiType,
            WikiType.TOPIC,
            parent
        ) { "No topic found to create a child topic with parentId $parentId" }

        val id = wikiSqlAdapter.createTopic(TopicEntity(null, topic, mutableListOf()))
        wikiSqlAdapter.updateTopic(
            parent.copy(childIds = parent.childIds.toMutableList().apply { add(id) })
        )

        return id
    }

    @Transactional
    override fun updateTopic(id: Int, topic: String, hasAccess: () -> Boolean): Int {
        hasAccess.checkAccess("Access denied for updating topic with id $id")

        val wikiType = wikiSqlAdapter.getWikiType(id)
        val topicEntity = wikiSqlAdapter.getTopic(id)

        verify(wikiType, WikiType.TOPIC, topicEntity) { "No topic found to update with id $id" }

        return wikiSqlAdapter.updateTopic(topicEntity.copy(topic = topic))
    }

    @Transactional
    override fun deleteTopic(id: Int, hasAccess: () -> Boolean) {
        hasAccess.checkAccess("Access denied for deleting topic for id $id")

        val wikiType = wikiSqlAdapter.getWikiType(id)
        val topic = wikiSqlAdapter.getTopic(id)
        val parent = wikiSqlAdapter.getTopicForChild(id)

        verify(wikiType, WikiType.TOPIC, topic) { "No topic found to delete with id $id" }
        verify(
            wikiType,
            WikiType.TOPIC,
            parent
        ) { "Can't delete topic with id $id because no parent topic found" }

        topic.childIds.forEach { wikiSqlAdapter.deleteEntry(it) }
        wikiSqlAdapter.updateTopic(
            parent.copy(childIds = parent.childIds.toMutableList().apply { remove(id) })
        )
        wikiSqlAdapter.deleteTopic(id)
    }

    override fun getEntry(id: Int, hasAccess: () -> Boolean): EntryEntity? {
        hasAccess.checkAccess("Access denied for getting entry with id $id")
        return wikiSqlAdapter.getEntry(id)
    }

    @Transactional
    override fun createEntry(
        topicId: Int,
        headline: String,
        body: String,
        hasAccess: () -> Boolean
    ): Int {
        hasAccess.checkAccess("Access denied for creating entry with id $topicId")

        val wikiType = wikiSqlAdapter.getWikiType(topicId)
        val parent = wikiSqlAdapter.getTopic(topicId)

        verify(
            wikiType,
            WikiType.TOPIC,
            parent
        ) { "Try to create entry for an undefined topic with topicId $topicId" }

        val entryId = wikiSqlAdapter.createEntry(EntryEntity(null, headline, body))
        wikiSqlAdapter.updateTopic(
            parent.copy(childIds = parent.childIds.toMutableList().apply { add(entryId) })
        )

        return entryId
    }

    @Transactional
    override fun updateEntry(
        id: Int,
        headline: String,
        body: String,
        hasAccess: () -> Boolean
    ): Int {
        hasAccess.checkAccess("Access denied for updating entry with id $id")

        val wikiType = wikiSqlAdapter.getWikiType(id)
        val entity = wikiSqlAdapter.getEntry(id)

        verify(wikiType, WikiType.ENTRY, entity) { "No entry found to update with id $id" }

        return wikiSqlAdapter.updateEntry(entity.copy(headline = headline, htmlEntry = body))
    }

    @Transactional
    override fun deleteEntry(id: Int, hasAccess: () -> Boolean) {
        hasAccess.checkAccess("Access denied for deleting entry with id $id")

        val wikiType = wikiSqlAdapter.getWikiType(id)
        val parent = wikiSqlAdapter.getTopicForChild(id)

        verify(wikiType, WikiType.ENTRY, parent) { "No entry found to delete with id $id" }

        wikiSqlAdapter.updateTopic(
            parent.copy(childIds = parent.childIds.toMutableList().apply { remove(id) })
        )
        wikiSqlAdapter.deleteEntry(id)
    }

    override fun getWikiType(id: Int, hasAccess: () -> Boolean): WikiType? {
        hasAccess.checkAccess("Access denied for getting type with id $id")
        return wikiSqlAdapter.getWikiType(id)
    }

    override fun getTopicForChild(childId: Int, hasAccess: () -> Boolean): TopicEntity? {
        hasAccess.checkAccess("Access denied for getting parent topic with id $childId")
        return wikiSqlAdapter.getTopicForChild(childId)
    }
}